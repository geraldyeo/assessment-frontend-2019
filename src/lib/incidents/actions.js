export const CREATE_ONE_INCIDENT = 'CREATE_ONE_INCIDENT';

export const createOneIncident = (payload) => ({ type: CREATE_ONE_INCIDENT, payload });