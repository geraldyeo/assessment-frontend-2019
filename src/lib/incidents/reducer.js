import React, { createContext, useReducer } from 'react';
import { CREATE_ONE_INCIDENT } from './actions';

export const initialState = {
  incidents: [
    { key: 'bcg_1', title: 'First incident', assignee: 'Admin', status: 'Resolved' }
  ]
};

export const reducer = (state, { type, payload }) => {
  Object.freeze(state);

  /* eslint-disable no-fallthrough */
  switch(type) {
    case CREATE_ONE_INCIDENT: {
      if (payload) {
        console.log(payload);
        return {...state, incidents: [...state.incidents, payload]};
      }
    }
    default:
      return state;
  }
  /* eslint-enable no-fallthrough */
};

  export const IncidentContext = createContext(null);

  export const IncidentProvider = ({ children }) => {
    const [state, dispatch] = useReducer(reducer, initialState);
    return (
      <IncidentContext.Provider value={{ state, dispatch }}>
        {children}
      </IncidentContext.Provider>
    );
  }