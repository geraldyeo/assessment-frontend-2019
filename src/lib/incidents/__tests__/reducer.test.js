import {createOneIncident} from '../actions';
import {initialState, reducer} from '../reducer';

describe('Incident Reducer', () => {
  test('calling reducer with unknown action returns current state', () => {
    expect(reducer(initialState, {type: 'UNKOWN'})).toBe(initialState);
  });

  test('calling reducer with createOneIncident action adds to state', () => {
    const payload = { key: 'bcg_2', title: 'Second incident', assignee: 'Admin', status: false };
    expect(reducer(initialState, createOneIncident(payload))).toMatchSnapshot();
  });
});