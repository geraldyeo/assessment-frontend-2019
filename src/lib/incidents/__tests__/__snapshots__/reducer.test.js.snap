// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Incident Reducer calling reducer with createOneIncident action adds to state 1`] = `
Object {
  "incidents": Array [
    Object {
      "assignee": "Admin",
      "key": "bcg_1",
      "status": true,
      "title": "First incident",
    },
    Object {
      "assignee": "Admin",
      "key": "bcg_2",
      "status": false,
      "title": "Second incident",
    },
  ],
}
`;
